### PR TITLE
[Download] Download module exception correction

### PR DESCRIPTION
**Pull request by apietraszek**
Correcting exceptions for cancel, pause and resume methods.
NotFoundError and TypeMismatchError exceptions are removed from above functions.

Abandon method added.
ABANDONED value added to DownaloadState enum.

[ACR] http://suprem.sec.samsung.net/jira/browse/TWDAPI-222

Signed-off-by: Arkadiusz Pietraszek <a.pietraszek@partner.samsung.com>

### CODE REVIEW

**Comment by mijin85cho** [comment]
Hello, @surbhig12 
I hope you to review the file only under guide folder, the docs/application/web/guides/connectivity/download.md .

For other web api files, you don't need to.

File: docs/application/web/api/5.5/device_api/mobile/tizen/download.html [hunk 1]
```diff
@@ -520,13 +524,49 @@ <h4>Methods</h4>
...
+ The ID of the ongoing download operation to abandon.
+                </li>
+        </ul>
+</div>
+<div class="exceptionlist">
+<p><span class="except">Exceptions:</span></p>
+          <ul class="exception"><li>WebAPIException<ul>
 <li class="list"><p>
- with error type InvalidValuesError, if any of the input parameters contain an invalid value.
+ with error type InvalidValuesError, if any of the input parameters contains an invalid value.
```
[the remark this answers is not in the document]
Comment (apietraszek) [reply docs/application/web/api/5.5/device_api/mobile/tizen/download.html]: Done

File: docs/application/web/guides/connectivity/download.md [hunk 2]
```diff
@@ -8,7 +8,7 @@ The main features of the Download API include:
 
 - Managing downloads
 
-  You can [start, pause, resume, and cancel a download](#managing-downloads) of content using the `DownloadManager` interface (in [mobile](../../api/latest/device_api/mobile/tizen/download.html#DownloadManager), [wearable](../../api/latest/device_api/wearable/tizen/download.html#DownloadManager), and [TV](../../api/latest/device_api/tv/tizen/download.html#DownloadManager) applications).
+  You can [start, pause, resume, cancel and abandon a download](#managing-downloads) of content using the `DownloadManager` interface (in [mobile](../../api/latest/device_api/mobile/tizen/download.html#DownloadManager), [wearable](../../api/latest/device_api/wearable/tizen/download.html#DownloadManager), and [TV](../../api/latest/device_api/tv/tizen/download.html#DownloadManager) applications).
```
[the remark this answers is not in the document]
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: Done

File: docs/application/web/guides/connectivity/download.md [hunk 3]
```diff
@@ -85,25 +85,43 @@ To provide the user access to Internet resources, you must learn how to manage d
 
    The `start()` method returns a unique identifier for the download operation.
 
-5. During the download:
+   You can check status of the download operation by calling getState() method with the download ID:
```
[the remark this answers is not in the document]
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: Done

File: docs/application/web/guides/connectivity/download.md [hunk 4]
```diff
@@ -85,25 +85,43 @@ To provide the user access to Internet resources, you must learn how to manage d
...
+
+      ```
+      tizen.download.getState(downloadId);
+      ```
+
+5. During the download operation, you can modify it's state.
```
[the remark this answers is not in the document]
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: Done

File: docs/application/web/guides/connectivity/download.md [hunk 5]
```diff
@@ -85,25 +85,43 @@ To provide the user access to Internet resources, you must learn how to manage d
...
+
+    Please note, that download operation can become completed almost instantaneously, depending on download size and connection speed.
```
[the remark this answers is not in the document]
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: Done

File: docs/application/web/guides/connectivity/download.md [hunk 6]
```diff
@@ -85,25 +85,43 @@ To provide the user access to Internet resources, you must learn how to manage d
...
 
    1. To pause the download, use the `pause()` method with the download ID:
 
       ```
       tizen.download.pause(downloadId);
       ```
+        Calling pause method will result in PAUSED download state only if previous state was DOWNLOADING or QUEUED
```
[the remark this answers is not in the document]
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: Done

File: docs/application/web/guides/connectivity/download.md [hunk 7]
```diff
@@ -85,25 +85,43 @@ To provide the user access to Internet resources, you must learn how to manage d
...
 
    2. To resume the download, use the `resume()` method with the download ID:
 
       ```
       tizen.download.resume(downloadId);
       ```
+        Calling resume method will result in DOWNLOADING or QUEUED download state only if previous state was PAUSED, CANCELED or FAILED
```
[the remark this answers is not in the document]
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: Done

File: docs/application/web/guides/connectivity/download.md [hunk 8]
```diff
@@ -85,25 +85,43 @@ To provide the user access to Internet resources, you must learn how to manage d
...
 
    3. To cancel the download, use the `cancel()` method with the download ID:
 
       ```
       tizen.download.cancel(downloadId);
       ```
+        Calling cancel method will result in CANCELED download state only if previous state was DOWNLOADING or QUEUED
```
[the remark this answers is not in the document]
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: Done

File: docs/application/web/guides/connectivity/download.md [hunk 9]
```diff
@@ -85,25 +85,43 @@ To provide the user access to Internet resources, you must learn how to manage d
...
+
+   4. To abandon the download, use the `abandon()` method with the download ID:
+
+      ```
+      tizen.download.abandon(downloadId);
+      ```
+        The abandoned download operation will have it's resources freed making it impossible to perform any further operations on this download operation.
```
[the remark this answers is not in the document]
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: Done

File: docs/application/web/guides/connectivity/download.md [hunk 10]
```diff
@@ -138,7 +156,7 @@ To provide the user access to Internet resources, you must learn how to check th
    ```
 
 ## Related Information
-* Dependencies   
+* Dependencies
```
[the remark this answers is not in the document]
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: Done

File: docs/application/web/guides/connectivity/download.md [hunk 11]
```diff
@@ -85,25 +85,45 @@ To provide the user access to Internet resources, you must learn how to manage d
...
       tizen.download.resume(downloadId);
       ```
+        Calling the resume() method will result in the DOWNLOADING or the QUEUED download state only if the previous state was PAUSED, CANCELLED or FAILED.
 
    3. To cancel the download, use the `cancel()` method with the download ID:
 
       ```
       tizen.download.cancel(downloadId);
       ```
+        Calling the cancel() method will result in the CANCELED download state only if the previous state was DOWNLOADING or QUEUED.
```
Comment (nawab-samsung) [inline docs/application/web/guides/connectivity/download.md]: @apietraszek 

Request you to change:   
CANCELED --> CANCELLED

File: docs/application/web/guides/connectivity/download.md [hunk 12]
```diff
@@ -85,25 +85,45 @@ To provide the user access to Internet resources, you must learn how to manage d
...
+
+      ```
+      tizen.download.getState(downloadId);
+      ```
+
+5. During the download operation, you can modify the download state.
+
+    >**Note**
+    >
+    >Download operation can become completed almost instantaneously, depending on download size and connection speed.
```
Comment (nawab-samsung) [inline docs/application/web/guides/connectivity/download.md]: @apietraszek 
Kindly change: 
 >Download operation can become completed almost instantaneously, depending on download size and connection speed.

to 

>The download operation gets completed almost instantaneously, depending on download size and connection speed.

File: docs/application/web/guides/connectivity/download.md (hunk 11 above)
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: I have changed others to CANCELED, that is the proper enumerator value.

File: docs/application/web/guides/connectivity/download.md (hunk 12 above)
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: Done

File: docs/application/web/guides/connectivity/download.md [hunk 13]
```diff
@@ -85,25 +85,45 @@ To provide the user access to Internet resources, you must learn how to manage d
 
    The `start()` method returns a unique identifier for the download operation.
 
-5. During the download:
+   You can check the status of the download operation by calling the getState() method with the download ID:
```
[the remark this answers is not in the document]
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: Done

File: docs/application/web/guides/connectivity/download.md [hunk 14]
```diff
@@ -85,25 +85,45 @@ To provide the user access to Internet resources, you must learn how to manage d
...
+
+      ```
+      tizen.download.getState(downloadId);
+      ```
+
+5. During the download operation, you can modify the download state.
+
+    >**Note**
```
[the remark this answers is not in the document]
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: Done

File: docs/application/web/guides/connectivity/download.md (hunk 11 above)
Comment (nawab-samsung) [reply docs/application/web/guides/connectivity/download.md]: @apietraszek 

thanks for the clarification, 
Line 112, please change it to CANCELED

Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: It's CANCELED in whole file now.

File: docs/application/web/guides/connectivity/download.md [hunk 15]
```diff
@@ -85,25 +85,45 @@ To provide the user access to Internet resources, you must learn how to manage d
...
+    >
+    >The download operation gets completed almost instantaneously, depending on download size and connection speed.
 
    1. To pause the download, use the `pause()` method with the download ID:
 
       ```
       tizen.download.pause(downloadId);
       ```
+        Calling the `pause()` method will result in the PAUSED download state only if the previous state was DOWNLOADING or QUEUED.
```
[the remark this answers is not in the document]
Comment (nawab-samsung) [reply docs/application/web/guides/connectivity/download.md]: @mijin85cho Thanks for your suggestion. 
Request you to give your valuable opinion. 
How about using the following sentence?

The `pause()` method results in the PAUSED download state only if the previous states were  DOWNLOADING or QUEUED.

Here I am keeping Subject and predicate in agreement with the verb.

File: docs/application/web/guides/connectivity/download.md [hunk 16]
```diff
@@ -85,25 +85,45 @@ To provide the user access to Internet resources, you must learn how to manage d
...
 
    2. To resume the download, use the `resume()` method with the download ID:
 
       ```
       tizen.download.resume(downloadId);
       ```
+        Calling the `resume()` method will result in the DOWNLOADING or the QUEUED download state only if the previous state was PAUSED, CANCELED or FAILED.
```
[the remark this answers is not in the document]
Comment (nawab-samsung) [reply docs/application/web/guides/connectivity/download.md]: @mijin85cho Thanks for your suggestion.
Request you to give your valuable opinion.
How about using the following sentence?

The `resume()` method results in the DOWNLOADING or the QUEUED download state only if the previous states were PAUSED, CANCELED or FAILED.

Here I am keeping Subject and predicate in agreement with the verb.

File: docs/application/web/guides/connectivity/download.md [hunk 17]
```diff
@@ -85,25 +85,45 @@ To provide the user access to Internet resources, you must learn how to manage d
...
       tizen.download.cancel(downloadId);
       ```
+        Calling the `cancel()` method will result in the CANCELED download state only if the previous state was DOWNLOADING or QUEUED.
+
+   4. To abandon the download, use the `abandon()` method with the download ID:
+
+      ```
+      tizen.download.abandon(downloadId);
+      ```
+        The abandoned download operation will have its resources freed up, making it impossible to perform any further operations on this download operation.
```
[the remark this answers is not in the document]
Comment (nawab-samsung) [reply docs/application/web/guides/connectivity/download.md]: @mijin85cho 
Thanks for your valuable suggestions. 
I will definitely take your suggestions as future tense must be avoided as much as possible. 

Reason: 

I kept the original format only because I was not sure that if abandoned download operation has the capability to free its resources and I did not want to rephrase without knowing.

The abandoned download operation will have its resources freed up, making it impossible to perform any further operations on this download operation.

Abandoned download operation may or may not have the capability to free its resources, but some other service may be called while the download is abandoning to free its resources.
So this made me to not to change the tense. 
Since this whole process depends on the abandonment of download. So only when the download will abandon, then this will happen, so the future tense.

File: docs/application/web/guides/connectivity/download.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (nawab-samsung) [reply docs/application/web/guides/connectivity/download.md]: Thanks @mijin85cho for your valuable suggestions: 

Request your opinion on the following sentence.: 
The `resume()` method results in the DOWNLOADING or the QUEUED download state only if the previous state were PAUSED, CANCELED or FAILED.

File: docs/application/web/guides/connectivity/download.md [hunk 18]
```diff
@@ -85,25 +85,45 @@ To provide the user access to Internet resources, you must learn how to manage d
...
       tizen.download.resume(downloadId);
       ```
+        Calling the `resume()` method will result in the DOWNLOADING or the QUEUED download state only if the previous state was PAUSED, CANCELED or FAILED.
 
    3. To cancel the download, use the `cancel()` method with the download ID:
 
       ```
       tizen.download.cancel(downloadId);
       ```
+        Calling the `cancel()` method will result in the CANCELED download state only if the previous state was DOWNLOADING or QUEUED.
```
[the remark this answers is not in the document]
Comment (nawab-samsung) [reply docs/application/web/guides/connectivity/download.md]: Thanks @mijin85cho for your valuable suggestions:

Request your opinion on the following sentence.:
The `cancel()` method results in the CANCELED download state only if the previous state were DOWNLOADING or QUEUED.

File: docs/application/web/guides/connectivity/download.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (nawab-samsung) [reply docs/application/web/guides/connectivity/download.md]: Thanks @mijin85cho for your valuable suggestions:

Request your opinion on the following sentence.:
The `pause()` method will result in the PAUSED download state only if the previous state were DOWNLOADING or QUEUED.

[the remark this answers is not in the document]
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: Download operation has one `state` variable. Its value can be changed. What I meant writing this sentence is following:

If current state is DOWNLOADING and `pause()` is called, state should change to PAUSED.
Also if current state is QUEUED and `pause()` is called, state should change to PAUSED.

I think that the sentence:

The `pause()` method results in the PAUSED download state only if the previous states were DOWNLOADING or QUEUED.

can be misunderstood that there is more than one state variable or maybe that the condition is met while state variable had its value set to DOWNLOADING or QUEUED in some point in time, but it needs to be in the moment of calling `pause()` method, for the method to take effect.

Please let me know what final form should this sentence take.

File: docs/application/web/guides/connectivity/download.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: When `abandon()` method is called with the id of the download operation, the method gives signal to free the operations resources and almost immediately the `abandon()` method is internally provided with feedback if this was done correctly, only after the feedback the `abandon()` method ends without rising exceptions.
So if `abandon()` method has been called and no exceptions was raised, then resources are already freed.

File: docs/application/web/guides/connectivity/download.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (nawab-samsung) [reply docs/application/web/guides/connectivity/download.md]: @apietraszek 
How about these sentences? 
Please verify and provide your opinion.
Line 105 - If the current state is DOWNLOADING or QUEUED, calling the `pause()` method changes the download state to PAUSED.

Line 112- If the current state is PAUSED, CANCELED, or FAILED, calling the `resume()` method changes the download state to DOWNLOADING or QUEUED.

Line 119- If the current state is DOWNLOADING or QUEUED, calling the `cancel()` method changes the download state to CANCELED.

File: docs/application/web/guides/connectivity/download.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (nawab-samsung) [reply docs/application/web/guides/connectivity/download.md]: @apietraszek 

Thanks for the details, I think @mijin85cho 's sentence is fine then.

The abandoned download operation frees its resources and stops further download operations. 

Kindly let me know about your opinion.

[the remark this answers is not in the document]
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: Done

File: docs/application/web/guides/connectivity/download.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: Thank You. I believe that will be unambiguous and in accordance with the facts.

Change done.

**Comment by mijin85cho** [comment]
Thanks for the review @nawab-samsung 
@apietraszek, for the update in api files, would you please get approval from one of team member?

File: docs/application/web/guides/connectivity/download.md [hunk 19]
```diff
@@ -85,25 +85,45 @@ To provide the user access to Internet resources, you must learn how to manage d
...
       tizen.download.cancel(downloadId);
       ```
+        If the current state is DOWNLOADING or QUEUED, calling the `cancel()` method changes the download state to CANCELED.
+
+   4. To abandon the download, use the `abandon()` method with the download ID:
+
+      ```
+      tizen.download.abandon(downloadId);
+      ```
+        The abandoned download operation frees its resources and stops further download operations.
```
Comment (pkosko) [inline docs/application/web/guides/connectivity/download.md]: I am not sure if the download operation frees its resources, there is a problem with subject of this sentence. Moreover, this comment suggest that 'abandon' stops download operationS - not only this one. I suggest rephrasing this comment to:
For the abandoned download operation, all its resources are released. After that, operation cannot be resumed.

File: docs/application/web/guides/connectivity/download.md [hunk 20]
```diff
@@ -85,25 +85,45 @@ To provide the user access to Internet resources, you must learn how to manage d
...
+
+      ```
+      tizen.download.getState(downloadId);
+      ```
+
+5. During the download operation, you can modify the download state.
+
+    >**Note**
+    >
+    >Depending on the download size and the connection speed, the download operation completes almost instantly.
```
Comment (pkosko) [inline docs/application/web/guides/connectivity/download.md]: ...  the download operation could complete almost instantly.  - sounds better for me.

File: docs/application/web/guides/connectivity/download.md [hunk 21]
```diff
@@ -138,7 +158,7 @@ To provide the user access to Internet resources, you must learn how to check th
    ```
 
 ## Related Information
-* Dependencies   
+- Dependencies
```
Comment (pkosko) [inline docs/application/web/guides/connectivity/download.md]: why did you change the enumeration type?

Comment (nawab-samsung) [reply docs/application/web/guides/connectivity/download.md]: @pkosko 

This enum change is as per our style guide. 
Kindly refer to [Styleguide](https://github.com/Samsung/tizen-docs/blob/master/styleguide/style.md#document-layout)

Also while parsing and converting our MD format to HTML, sometimes if we use any other symbol(+,*)  other than (-) conversion didn't go well sometimes.

File: docs/application/web/guides/connectivity/download.md (hunk 20 above)
Comment (nawab-samsung) [reply docs/application/web/guides/connectivity/download.md]: @pkosko  we avoid using future tenses as a part of thumb rule for Tizen documentation. That's why we must not use (could, would, will) or any word that represents future tense. 

Reference: [Google Style guide](https://developers.google.com/style/tense)

File: docs/application/web/guides/connectivity/download.md (hunk 19 above)
Comment (nawab-samsung) [reply docs/application/web/guides/connectivity/download.md]: @pkosko I too was sceptic about the same, refer to https://github.com/Samsung/tizen-docs/pull/827#discussion_r327441130 . 

Just a suggestion: 
How about using this? 

For the abandoned download operation since all its resources are already released thus, download operation cannot be resumed. 
@mijin85cho 
Please share your opinion on this.

File: docs/application/web/guides/connectivity/download.md (hunk 20 above)
Comment (pkosko) [reply docs/application/web/guides/connectivity/download.md]: @nawab-samsung then I suggest rephrase like below:

For downloading small files and high connection speed, the download operation completes almost instantly.

File: docs/application/web/guides/connectivity/download.md (hunk 19 above)
Comment (pkosko) [reply docs/application/web/guides/connectivity/download.md]: "For any abandoned download operation, the resources are already released. Thus the download operation cannot be resumed." I ok for me.

File: docs/application/web/guides/connectivity/download.md (hunk 21 above)
Comment (pkosko) [reply docs/application/web/guides/connectivity/download.md]: ok

File: docs/application/web/guides/connectivity/download.md (hunk 19 above)
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: Done

File: docs/application/web/guides/connectivity/download.md (hunk 20 above)
Comment (apietraszek) [reply docs/application/web/guides/connectivity/download.md]: After discussion with Piotr Koska we decided to change it to:

"For downloading small file with using high speed connection, the download operation completes almost instantly."

@nawab-samsung please let us know if that's ok.

Comment (nawab-samsung) [reply docs/application/web/guides/connectivity/download.md]: Slight change
 For downloading small files using the high-speed connection, the download operation completes almost instantly.
